File: warmup/01-CountingValleys.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the countingValleys function below.
    static int countingValleys(int n, String s) {
        int valleyCount = 0;
        int level = 0;
        int prevLevel = 0;
        
        for (int i = 0; i < s.length(); i++) {
            prevLevel = level;
            if (s.charAt(i) == 'U') {
                level += 1;
            } else {
                level -= 1;
            }
            
            if (level < 0 && prevLevel >= 0) {
                valleyCount++;
            }
        }
        
        return valleyCount;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        String s = scanner.nextLine();

        int result = countingValleys(n, s);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}

File: warmup/01-CountingValleys.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the countingValleys function below.
def countingValleys(n, s):
    height = 0
    inValley = False
    valleyCounter = 0
    for letter in s:
        if letter == 'U':
            height += 1
        elif letter == 'D':
            height -= 1
        
        if inValley == False and height < 0:
            valleyCounter += 1
            inValley = True
        elif inValley == True and height >= 0:
            inValley = False
    return valleyCounter
        

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    s = input()

    result = countingValleys(n, s)

    fptr.write(str(result) + '\n')

    fptr.close()

